torch/utils/data/dataloader.py

```diff
@@ -425,80 +491,77 @@ class _DataLoaderIter(object):
-
-        self.sample_iter = iter(self.batch_sampler)
-
-        base_seed = torch.LongTensor(1).random_().item()
-
-        if self.num_workers > 0:
-            self.worker_init_fn = loader.worker_init_fn
-            self.worker_queue_idx = 0
-            self.worker_result_queue = multiprocessing.Queue()
-            self.batches_outstanding = 0
-            self.worker_pids_set = False
-            self.shutdown = False
-            self.send_idx = 0
-            self.rcvd_idx = 0
-            self.reorder_dict = {}
-            self.done_event = multiprocessing.Event()
-
-            self.index_queues = []
-            self.workers = []
-            for i in range(self.num_workers):
-                index_queue = multiprocessing.Queue()
-                index_queue.cancel_join_thread()
-                w = multiprocessing.Process(
-                    target=_utils.worker._worker_loop,
-                    args=(self.dataset, index_queue,
-                          self.worker_result_queue, self.done_event,
-                          self.collate_fn, base_seed + i,
-                          self.worker_init_fn, i))
-                w.daemon = True
-                # NB: Process.start() actually take some time as it needs to
-                #     start a process and pass the arguments over via a pipe.
-                #     Therefore, we only add a worker to self.workers list after
-                #     it started, so that we do not call .join() if program dies
-                #     before it starts, and __del__ tries to join but will get:
-                #     AssertionError: can only join a started process.
-                w.start()
-                self.index_queues.append(index_queue)
-                self.workers.append(w)
-
-            if self.pin_memory:
-                self.data_queue = queue.Queue()
-                pin_memory_thread = threading.Thread(
-                    target=_utils.pin_memory._pin_memory_loop,
-                    args=(self.worker_result_queue, self.data_queue,
-                          torch.cuda.current_device(), self.done_event))
-                pin_memory_thread.daemon = True
-                pin_memory_thread.start()
-                # Similar to workers (see comment above), we only register
-                # pin_memory_thread once it is started.
-                self.pin_memory_thread = pin_memory_thread
-            else:
-                self.data_queue = self.worker_result_queue
-
-            _utils.signal_handling._set_worker_pids(id(self), tuple(w.pid for w in self.workers))
-            _utils.signal_handling._set_SIGCHLD_handler()
-            self.worker_pids_set = True
+        super(_MultiProcessingDataLoaderIter, self).__init__(loader)
+
+        assert self.num_workers > 0
+
+        self.worker_init_fn = loader.worker_init_fn
+        self.worker_queue_idx_cycle = itertools.cycle(range(self.num_workers))
+        self.worker_result_queue = multiprocessing.Queue()
```

Comment:
I'm wondering if this queue needs to have some max capacity? It seems like we might run into memory issues if we have a very large dataset and the workers produce work at a much faster than what the application can consume.

Reply:
there is an upper bound of number of outstanding jobs (2*num_workers), so no this does not need a max capacity.

Reply:
This is also not a new behavior introduced by this pr.